doc/developer/design/20250203_cluster_reliability.md

```diff
@@ -0,0 +1,224 @@
+
+## Solution Proposal
+
+<!--
+What is your preferred solution, and why have you chosen it over the
+alternatives? Start this section with a brief, high-level summary.
+
+This is your opportunity to clearly communicate your chosen design. For any
+design document, the appropriate level of technical details depends both on
+the target reviewers and the nature of the design that is being proposed.
+A good rule of thumb is that you should strive for the minimum level of
+detail that fully communicates the proposal to your reviewers. If you're
+unsure, reach out to your manager for help.
+
+Remember to document any dependencies that may need to break or change as a
+result of this work.
+-->
+
+User's needs aren't uniform; a solution working for one customer might be too limiting for another.
+We propose a set of mechanisms that offer a spectrum of solutions:
+* We add a RBAC permission `CREATE DATAFLOW`, which is required to create dataflows.
+  If the permission is absent, select and subscribe queries can only read from existing objects.
+* We add a RBAC permission `READ EXTERNAL`, which is required to read data from external storage (persist).
+  The permission controls both dataflows reading from external sources, and peeks that directly read from external storage.
+* We implement a mechanism to limit the resources of a dataflow serving a query with a probability.
+
+### `CREATE DATAFLOW` permission
+
+The `CREATE DATAFLOW` permission controls whether a query can render its own dataflow.
+Without the permission, Materialize will reject a query that it cannot otherwise fulfill.
+The permission applies to a cluster.
+
+Implementing this requires additional plumbing.
+At the moment, we evaluate the RBAC permissions before sequencing a plan.
+However, we only learn during sequencing whether an object needs to render a dataflow.
+
+### `READ EXTERNAL` permission
+
+The `READ EXTERNAL` permission controls whether a query can read from external storage.
+If a query can only be fulfilled by reading data from external storage, Materialize checks that the issuer has the permission on the target cluster.
+
+The permission would apply to all objects, i.e., indexes, materialized views, and queries.
+Similar considerations as with `CREATE DATAFLOW` regarding the implementation apply here, too.
+
+### Limiting a query's heap size
+
+Measuring heap size usage is inherently difficult as we cannot attribute all memory usage to a single query.
+Any operator can allocate and deallocate memory, and we do not want to (or cannot) instrument all operators.
+We can only measure instrumented dataflow operators, and with a time delay, which renders all solutions approximate.
+
+A solution needs to limit the heap size usage of a query to a configurable value, and enforce it with high probability.
+
+To measure the heap size of a query, we hook into the arrangement heap size infrastructure.
```

Comment:
Will this count consolidations in `Union`s or the cost of finishing? (I don't know if our sort implementation is in-place or allocating, for example.)